feat(ISV-5127): Add new Tekton task to Konflux to upload sbom to Atlas

ezopezo opened the pull request:
New Tekton task is added to Konflux for uploading SBOM-s to Atlas.
Upload supports both CycloneDX and SPDX, and be able to convert both formats to versions supported by Atlas before the upload (2.3 for SPDX, 1.4 for CycloneDX or lower).

mmalina commented on tasks/upload-sbom-to-atlas/README.md:
```diff
@@ -0,0 +1,16 @@
+
+| Name                      | Description                                                                                                                                                                                                                          | Optional | Default value                                                                 |
+|---------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------|-------------------------------------------------------------------------------|
+| sbomsDir                  | Directory containing SBOM files. The task will search for CycloneDX JSON SBOMs recursively in this directory and upload them all to Atlas. The path is relative to the 'sboms' workspace.                                            | true     | .                                                                             |
```
We have a convention of using Yes/No for the Optional column. So please change this to "Yes". This applies to the rest as well.

mmalina commented on tasks/upload-sbom-to-atlas/tests/mocks.sh:
```diff
@@ -0,0 +1,10 @@
+  # Output the call to stderr
+  echo "Mock curl called with:" "$@" >&2
+  echo -n "$@" >> "${workdir}/mock_curl.txt"
+  tail -c1 "${workdir}/mock_curl.txt" | read -r _ || printf "\n" >> "${workdir}/mock_curl.txt"
```
What's the purpose of this? Why not just print the new line in the echo above (i.e. remove `-n`)?

ezopezo replied:
This is leftover from my local testing, thank you for pointing it out!

mmalina commented on tasks/upload-sbom-to-atlas/upload-sbom-to-atlas.yaml:
```diff
@@ -0,0 +1,261 @@
+        fi
+  
+        shopt -s nullglob
+        sboms_to_upload=("$workdir"/*.json)
```
I don't know how this is supposed to work - how `/tekton/home` is preserved among the steps or if you even have access to write there. It might. But I would recommend using an `emptyDir` for sharing data among steps.

See e.g. here: https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/push-rpm-data-to-pyxis/push-rpm-data-to-pyxis.yaml#L37

mmalina commented on tasks/upload-sbom-to-atlas/upload-sbom-to-atlas.yaml:
```diff
@@ -0,0 +1,261 @@
+        
+          # Check if SBOM is CycloneDX or SPDX
+          if [ "$sbom_format" == "CycloneDX" ]; then
+            echo "Found CycloneDX SBOM: $file_relpath"
```
The same is already printed on line 123.

mmalina commented on tasks/upload-sbom-to-atlas/upload-sbom-to-atlas.yaml:
```diff
@@ -0,0 +1,261 @@
+            printf "%s" "cyclonedx-json" > "${sbom_path}.sbom_format"
+        
+          elif [ "$sbom_format" == "SPDX" ]; then
+            echo "Found SPDX SBOM: $file_relpath"
```
The same is already printed on line 123.

ezopezo replied:
Good catch, thanks!

mmalina commented on tasks/upload-sbom-to-atlas/upload-sbom-to-atlas.yaml:
```diff
@@ -0,0 +1,261 @@
+          sbom_format=$(jq -r 'if .bomFormat then .bomFormat else if .spdxVersion then "SPDX" else empty end end' "$filepath")
+        
+          # If sbom_format is empty, it's not a valid SBOM (CycloneDX or SPDX)
+          if [ -z "$sbom_format" ]; then
```
This is already covered by the `else` on line 149. So either this can be removed. Or maybe it's more readable to have an explicit check at the top, in which case this could cover the other case as well. Either, having two places where we say `not a valid SBOM` and continue the loop seems suboptimal.

mmalina commented on tasks/upload-sbom-to-atlas/upload-sbom-to-atlas.yaml:
```diff
@@ -0,0 +1,261 @@
+
+    # Needs syft, which is in a different image => has to be a separate step
+    - name: convert-sboms-if-needed
+      image: registry.redhat.io/rh-syft-tech-preview/syft-rhel9:1.0.1@sha256:27c268d678103a27b6964c2cd5169040941b7304d0078f9727789ffb8ffba370
```
This is problematic because this needs a token to pull and so github ci won't be able to run the tests. Maybe we should add syft to the utils image?

ezopezo replied:
Sure no problem, I just created PR [here](https://github.com/konflux-ci/release-service-utils/pull/264).

johnbieren commented:
/ok-to-test

mmalina commented:
@ezopezo please rebase again.

ezopezo committed:
feat(ISV-5127): add new Tekton task to Konflux to upload sboms to Atlas

This task is intended to gather SBOM files
from specified directory, convert them to
versions supported by Atlas and upload
them to Atlas.
Task supports SPDX and CycloneDX formats.

Signed-off-by: Erik Mravec <emravec@redhat.com>
